### DISCUSSION TITLE
Use system provided MPI for depedencies

### DISCUSSION THREAD

**Discussion by moritz-h** [microsoft/vcpkg#26134] [Q&A] [answered]
Hello,
are there best practices or recommendations, how we can use a system provided MPI installation as alternative to the `mpi` port?

For context, we are trying to use vcpkg for dependency management in an application which uses MPI, but also uses dependencies which use MPI (i.e. `vtk-m`, in addition to a few self written ports.) Currently using the default `mpi` port is working on a development machine, but we are trying to investigate if this setup will work on systems where only a specific pre-installed MPI version is available and the custom built openmpi or msmpi coming from vcpkg will not work (i.e. on some HPC systems).

A rough high-level idea would be to maybe do something like an overlay port for the `mpi` port, which under the hood will do something like a wrapper to the generic find MPI scripts build in to CMake. But before trying to do something like this, is there a solution or best practices available how to use system libraries with vcpkg (and also for vcpkg managed dependencies)? Or have you any other recommendations or experience how this could be solved?

**Comment by dg0yt** [comment, marked as answer]
An *empty* overlay port is a good way to deal with this.
If you have to handle multiple ports, you may also organize them in a "registry".

**Reply by moritz-h** [reply]
Thanks for the answer, but my question was meant in a different direction. Using overlay ports in general is clear. I wanted to know if it is possible to more or less workaround the vcpkg principle of building everything on its own from scratch. So my setup is, I want to build a port which depends on `mpi`. But now, some machines require their own specific MPI version for that machine. The genereal vcpkg solution would be to write an overlay port for the mpi package with that specific version. The problem is, I don't have sources or binaries for that specific MPI version to write a overlay port the regular way. Instead I want to just use the MPI installed on the system as I would do if using a system depdency without vcpkg, but in addition I must somehow still provide a vcpkg `mpi` port as depedency to other ports.

A rough idea would be to provide something like a dummy overlay port for `mpi` which pasically should only passthrough what could be found on the system anyway. My question would be, is that possible? If yes, is this a good idea or are there other recommendations how to solve the above descibed situation.

**Reply by Neumann-A** [reply]
vcpkg/cmake needs one mpi implementation to link against. If you use a mpi as a shared library it shouldn't matter which implementation you use (.dll/.so names have to be the same though) since mpi is a c library with a well defined API. As such it shouldn't matter which one you link in the build of other libraries as long as the executable you build is able to find a shared mpi library. 

If you do that however the version of the mpi library used needs to be the minimal version you are expected to use later on so that you cannot accidentally use mpi functionality added by later mpi versions.

**Reply by moritz-h** [reply]
I now found what I was looking for, basically the (now deprecated) option `VCPKG_PREFER_SYSTEM_LIBS` https://github.com/microsoft/vcpkg/blob/master/docs/users/buildsystems/cmake-integration.md#vcpkg_prefer_system_libs

But the now described alternative is exactly the answer I was looking for. The answer to how to write a "dummy port" to provide the vcpkg dependency to other ports while actually falling back to system libraries is just use an empty overlay port.

**Reply by dg0yt** [reply]
`VCPKG_PREFER_SYSTEM_LIBS` will hurt you, sooner or later. Don't use it. Don't recommend it here.

I will insert "empty" before "overlay port" to my answer, because it was really meant like this.

**Reply by moritz-h** [reply]
Yes, I do not intent to use `VCPKG_PREFER_SYSTEM_LIBS` just the name describes in short what I was meaning and the docs there recommend the empty port as alternative.
With the word "empty" the initial answer now makes a lot more sense, so marked it as answer now.
